File: src/vocab.ts
/**
 * Vocabulary loading, rendering, tokenizing, and the option-masking rules.
 */

import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

export const END = "<end>";
export const END_DESCRIPTION = "nothing should follow; the text is finished";
const PUNCTUATION = new Set([".", ",", "?", "!", ":", ";", "...", "-"]);

const PACKAGE_ROOT = join(dirname(fileURLToPath(import.meta.url)), "..");

/** Resolve a file inside the shipped `data/` directory. */
export function dataPath(...parts: string[]): string {
  return join(PACKAGE_ROOT, "data", ...parts);
}

export const DEFAULT_VOCAB = (): string => dataPath("vocab_en.txt");
export const DEFAULT_CORPUS = (): string => dataPath("corpus_en.txt");
export const DEFAULT_HELDOUT = (): string => dataPath("heldout_en.txt");

export class Vocab {
  readonly tokens: readonly string[];
  private readonly tokenSet: Set<string>;
  private readonly widest: number;

  constructor(tokens: readonly string[]) {
    if (tokens.length > 255) {
      throw new Error(`${tokens.length} options supplied; a Choice accepts at most 255`);
    }
    this.tokens = tokens;
    this.tokenSet = new Set(tokens);
    this.widest = tokens.reduce((widest, token) => Math.max(widest, token.split(" ").length), 1);
  }

  static load(path?: string): Vocab {
    const source = path ?? DEFAULT_VOCAB();
    const tokens = readFileSync(source, "utf8")
      .split("\n")
      .map((line) => line.trim())
      .filter((line) => line !== "" && !line.startsWith("#"));
    try {
      return new Vocab(tokens);
    } catch (error) {
      throw new Error(`${source}: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * The criteria map for a Choice question. Descriptions stay null where the name
   * speaks for itself: every option costs tokens on every call.
   */
  options(): Record<string, string | null> {
    const options: Record<string, string | null> = {};
    for (const token of this.tokens) {
      options[token] = token === END ? END_DESCRIPTION : null;
    }
    return options;
  }

  /** Join tokens into readable text. Punctuation sticks to the previous token. */
  render(tokens: readonly string[]): string {
    let out = "";
    for (const token of tokens) {
      if (out === "") out = token;
      else if (PUNCTUATION.has(token)) out += token;
      else out += ` ${token}`;
    }
    return out;
  }

  /** Longest-match tokenize. null marks a word with no matching option. */
  tokenize(text: string): (string | null)[] {
    const words = text.split(/\s+/).filter((word) => word !== "");
    const tokens: (string | null)[] = [];
    let index = 0;
    while (index < words.length) {
      let matched = false;
      for (let width = this.widest; width >= 1; width -= 1) {
        const candidate = words.slice(index, index + width).join(" ");
        if (candidate === "") continue;
        const canonical = candidate.toLowerCase();
        if (this.tokenSet.has(canonical)) {
          tokens.push(canonical);
          index += width;
          matched = true;
          break;
        }
        if (this.tokenSet.has(candidate)) {
          tokens.push(candidate);
          index += width;
          matched = true;
          break;
        }
      }
      if (!matched) {
        tokens.push(null);
        index += 1;
      }
    }
    return tokens;
  }

  /**
   * Options that survive the anti-degeneracy mask.
   *
   * Measured failure mode: given text ending in a space or a period, Jev puts most of
   * its mass on repeating that character or word. Masking the repeat in code, then
   * renormalizing, is what breaks the loop.
   */
  allowed(tokens: readonly string[], minTokens = 0, banRepeatBigrams = true): Set<string> {
    const allowed = new Set(this.tokens);
    if (tokens.length < minTokens) allowed.delete(END);
    const last = tokens.at(-1);
    if (last !== undefined) allowed.delete(last);
    if (banRepeatBigrams && tokens.length >= 2) {
      const seen = new Set<string>();
      for (let index = 0; index < tokens.length - 1; index += 1) {
        seen.add(`${tokens[index]}\u0000${tokens[index + 1]}`);
      }
      for (const token of allowed) {
        if (last !== undefined && seen.has(`${last}\u0000${token}`)) allowed.delete(token);
      }
    }
    return allowed;
  }
}

/** Drop masked options and rescale. Falls back to the raw distribution if empty. */
export function renormalize(
  probabilities: Record<string, number>,
  allowed?: ReadonlySet<string> | null,
): Record<string, number> {
  if (allowed === undefined || allowed === null) return { ...probabilities };
  const kept: Record<string, number> = {};
  let total = 0;
  for (const [token, probability] of Object.entries(probabilities)) {
    if (allowed.has(token) && probability > 0) {
      kept[token] = probability;
      total += probability;
    }
  }
  if (total <= 0) return { ...probabilities };
  for (const token of Object.keys(kept)) {
    kept[token] = (kept[token] ?? 0) / total;
  }
  return kept;
}

File: python/jev_cli/vocab.py
"""Vocabulary loading, rendering, tokenizing, and the option-masking rules."""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path

DATA_DIR = Path(__file__).parent / "data"
DEFAULT_VOCAB = DATA_DIR / "vocab_en.txt"
DEFAULT_CORPUS = DATA_DIR / "corpus_en.txt"
END = "<end>"
NONE = "<none>"
PUNCTUATION = {".", ",", "?", "!", ":", ";", "...", "-"}
END_DESCRIPTION = "nothing should follow; the text is finished"
NONE_DESCRIPTION = "the next token is not among the options above"


@dataclass(frozen=True)
class Vocab:
    tokens: tuple[str, ...]

    @classmethod
    def load(cls, path: str | Path | None = None) -> "Vocab":
        source = Path(path) if path else DEFAULT_VOCAB
        tokens = [
            line.strip()
            for line in source.read_text().splitlines()
            if line.strip() and not line.lstrip().startswith("#")
        ]
        if len(tokens) > 255:
            raise ValueError(f"{source} has {len(tokens)} options; a Choice accepts at most 255")
        return cls(tuple(tokens))

    def options(self, allowed: set[str] | None = None) -> dict[str, str | None]:
        """The criteria map for a Choice question. Keep descriptions null where the
        name speaks for itself: every option costs tokens on every call.

        Pass `allowed` to offer only those tokens. The option count is the cost of the
        call in both directions: 229 options measured 1706 in and 1638 out tokens, an
        11-option map measured 364 in and 98 out on the same state.
        """
        tokens = self.tokens if allowed is None else tuple(t for t in self.tokens if t in allowed)
        return {token: (END_DESCRIPTION if token == END else None) for token in tokens}

    def render(self, tokens: list[str] | tuple[str, ...]) -> str:
        """Join tokens into readable text. Punctuation sticks to the previous token."""
        out = ""
        for token in tokens:
            if not out:
                out = token
            elif token in PUNCTUATION:
                out += token
            else:
                out += " " + token
        return out

    def tokenize(self, text: str) -> list[str | None]:
        """Longest-match tokenize against this vocabulary. None marks a word with no option."""
        words = text.split()
        tokens: list[str | None] = []
        index = 0
        widest = max(len(token.split()) for token in self.tokens)
        while index < len(words):
            for width in range(widest, 0, -1):
                candidate = " ".join(words[index : index + width])
                canonical = candidate.lower()
                if canonical in self.tokens:
                    tokens.append(canonical)
                    index += width
                    break
                if candidate in self.tokens:
                    tokens.append(candidate)
                    index += width
                    break
            else:
                tokens.append(None)
                index += 1
        return tokens

    def allowed(
        self,
        tokens: list[str],
        min_tokens: int = 0,
        ban_repeat_bigrams: bool = True,
    ) -> set[str]:
        """Options that survive the anti-degeneracy mask.

        Measured failure mode: given a text ending in a space or a period, Jev puts
        most of its mass on repeating that character or word. Masking the repeat in
        code, then renormalizing, is what breaks the loop.
        """
        allowed = set(self.tokens)
        if len(tokens) < min_tokens:
            allowed.discard(END)
        if tokens:
            allowed.discard(tokens[-1])
        if ban_repeat_bigrams and len(tokens) >= 2:
            seen = set(zip(tokens, tokens[1:]))
            allowed = {token for token in allowed if (tokens[-1], token) not in seen}
        return allowed


def renormalize(probs: dict[str, float], allowed: set[str] | None) -> dict[str, float]:
    """Drop masked options and rescale. Falls back to the raw distribution if empty."""
    if allowed is None:
        return dict(probs)
    kept = {token: p for token, p in probs.items() if token in allowed and p > 0.0}
    total = sum(kept.values())
    if total <= 0.0:
        return dict(probs)
    return {token: p / total for token, p in kept.items()}
